src/base/sysinfo.cc

```diff
@@ -206,6 +206,110 @@ extern "C" {
   }
 }
 
+// A set of primitives to trick with the environemnt variables
+// to mark child processes.
+// As gperftools is commonly loaded dynamically using LD_PRELOAD,
+// users are leveraging environment variables to communicate the parameters
+// to the corresponding gperftools library.
+// Specifically, a set of environment variables are used to specify the file
+// paths to be used during the operation (i.e. output profile filename).
+// When the application is using fork() to spawn more processes, it is important
+// to avoid conflicts where both the parent and a child area using the same
+// filename.
+// In order to do so, the parent reads the user-defined environment and toggles
+// the MSB of the first byte in it's value to indicate that all other processes
+// using this variable are children (and thus) must append a unique identifier
+// (PID) as a suffix to the corresponsing filename.
+static inline void EnvChildSignSet(char *enval)
+{
+  enval[0] &= ~128;
+}
+
+static inline void EnvChildSignDrop(char *enval)
+{
+  enval[0] &= ~128;
+}
+
+static inline bool EnvChildSignIsOn(char *enval)
+{
+  return (enval[0] & 128);
+}
+
+// HPC environment auto-detection
+// For HPC applications (MPI, OpenSHMEM, etc), it is typical for multiple
+// processes not engaged in parent-child relations to be executed on the
+// same host.
+// In order to enable gperftools to analyze them, these processes need to be
+// assigned individual file paths for the files being used.
+// The function below is trying to discover well-known HPC environments and
+// take advantage of that environment to generate meaningful profile filenames
+bool DetectHPCEnvironment(char *suffix, size_t max_size)
+{
+  char* enval;
+
+  // Check for the PMIx environment
+  enval = getenv("PMIX_RANK");
+  if (enval) {
+    /* PMIx exposes the rank that is convenient to be used */
+    snprintf(suffix, max_size, ".pmixrank-%s", enval);
+    return true;
+  }
+
+  // Check for the Slurm environment
+  enval = getenv("SLURM_JOBID");
+  if (enval) {
+    enval = getenv("SLURM_PROCID");
+    if (enval) {
+      snprintf(suffix, max_size, ".slurmid-%s", enval);
+      return true;
+    }
+    goto use_getpid;
+  }
+
+  /* Check for Open MPI environment */
+  enval = getenv("OMPI_HOME");
+  if (enval) {
+    goto use_getpid;
+  }
+
+  // TODO: Detect other environments - MPICH
+
+  // No HPC environment was detected
+  return false;
+
+use_getpid:
+  snprintf(suffix, max_size, ".pid-%d", getpid());
+  return true;
+}
+
+void SetUniquePathSuffix(char *suffix, size_t max_size, bool is_child)
```

Comment:
So I see some duplication between here and above. Is it worth just having detect hpc thingy do the detect part and this function do the suffix part ?

Reply:
would be nice if you refactor so that no goto use_pid is necessary. Feels like it should be possible